File: 1296.divide-array-in-sets-of-k-consecutive-numbers.cpp
/*
 * @lc app=leetcode id=1296 lang=cpp
 *
 * [1296] Divide Array in Sets of K Consecutive Numbers
 */

// @lc code=start
class Solution {
public:
    // Same as Q846
    // T: O(n + m * log(m)) amortized (each number only used once), m := the number of unique value
    // S: O(n)
    bool isPossibleDivide(vector<int>& nums, int k) {
        if (nums.size() % k != 0) return false; 

        map<int, int> cnts;
        for (const int& x : nums) {
            ++cnts[x];
        }

        // For each unique value, if some of its instances have not been used in groups,
        // let these instances be the first numbers of the new groups, and check if we 
        // have enough instances of consecutive values to complete the new groups. 
        for (const auto& [x, x_cnt] : cnts) {
            if (x_cnt > 0) {
                for (int y = x + 1; y < x + k; ++y) {
                    auto p = cnts.find(y);
                    if (p == cnts.end() || p->second < x_cnt) {
                        return false;
                    }
                    p->second -= x_cnt;
                }
            }
        }

        return true;
    }
};
// @lc code=end

File: 1296.divide-array-in-sets-of-k-consecutive-numbers.py
#
# @lc app=leetcode id=1296 lang=python3
#
# [1296] Divide Array in Sets of K Consecutive Numbers
#
from collections import Counter
from typing import *

# @lc code=start
class Solution:
    # hash map + sorting + greedy
    # Same as Q846
    # T: O(n + m * log(m)) amortized, m := the number of unique numbers
    # S: O(m)
    def isPossibleDivide(self, nums: List[int], k: int) -> bool:
        if len(nums) % k:
            return False
        
        cnts = Counter(nums)
        
        for x in sorted(cnts.keys()):
            if cnts[x] > 0:
                required = cnts[x]

                # Note that k may not contribute to time complexity.
                # https://leetcode.com/problems/divide-array-in-sets-of-k-consecutive-numbers/discuss/470238/JavaC++Python-Exactly-Same-as-846.-Hand-of-Straights/1001094
                for y in range(x + 1, x + k):
                    if y not in cnts or cnts[y] < required:
                        return False

                    cnts[y] -= required   # IMPORTANT: consume next numbers
        
        return True
        
# @lc code=end

